resource/resource.go

```diff
@@ -10,16 +10,44 @@ import (
 	"strings"
 
 	gabs "github.com/Jeffail/gabs/v2"
+	"github.com/conduktor/ctl/orderedjson"
+	"github.com/conduktor/ctl/printutils"
 	yamlJson "github.com/ghodss/yaml"
 	yaml "gopkg.in/yaml.v3"
 )
 
 type Resource struct {
 	Json     []byte
 	Kind     string
-	Name     string
+	Name     string //duplicate data from Metadata.Name extracted for convenience
 	Version  string
-	Metadata map[string]interface{} `json:"-"`
+	Metadata map[string]interface{}
+	Spec     map[string]interface{}
+}
+
+func (r Resource) MarshalJSON() ([]byte, error) {
+	return r.Json, nil
+}
+
+func (r *Resource) UnmarshalJSON(data []byte) error {
+	var forParsingStruct forParsingStruct
+	err := json.Unmarshal(data, &forParsingStruct)
+	if err != nil {
+		return err
+	}
+
+	name, err := extractKeyFromMetadataMap(forParsingStruct.Metadata, "name")
+	if err != nil {
+		return err
+	}
+
+	r.Json = data
+	r.Kind = forParsingStruct.Kind
+	r.Name = name
+	r.Version = forParsingStruct.ApiVersion
+	r.Metadata = forParsingStruct.Metadata
+	r.Spec = forParsingStruct.Spec
+	return expendIncludeFiles(r)
```

Comment:
You mute current resource, not create and return a new one ? 
I would expect like Marshall that return the current resource as json []byte the unmarshall be more a constructor fromJsonByteArray or something

Reply:
Or maybe it's the "standard" go way of doing things when calling `json.Unmarshal(data, var)` In that case might need to be constant with yaml that have a FromYAMLByte constructor

Reply:
That is how go want me to define `UnmarshalJSON`